File: medium/2130. Maximum Twin Sum of a Linked List/2130. Maximum Twin Sum of a Linked List.go
package main

/**
 * Definition for singly-linked list.
 */

type ListNode struct {
	Val  int
	Next *ListNode
}

func pairSum(head *ListNode) int {
	// nums := []int{}
	// curr := head
	// for curr != nil {
	// 	nums = append(nums, curr.Val)
	// 	curr = curr.Next
	// }

	// mid := len(nums) / 2
	// curr = head
	// res := 0

	// for mid != 0 {
	// 	res = max(res, curr.Val+nums[len(nums)])
	// 	nums = nums[:len(nums)-1]
	// 	curr = curr.Next
	// 	mid -= 1
	// }

	// return res

	slow := head
	fast := head
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}

	var prev *ListNode
	curr := slow
	for curr != nil {
		next := curr.Next
		curr.Next = prev
		prev = curr
		curr = next
	}

	res := 0

	for prev != nil {
		res = max(res, prev.Val+head.Val)
		head = head.Next
		prev = prev.Next
	}

	return res
}

File: medium/2130. Maximum Twin Sum of a Linked List/2130. Maximum Twin Sum of a Linked List.py
from collections import deque
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        nums = deque()
        curr = head

        while curr:
            nums.append(curr.val)
            curr = curr.next

        mid = len(nums) // 2
        curr = head
        res = 0

        for _ in range(mid):
            if curr:
                res = max(res, curr.val + nums.popleft())
                curr = curr.next

        return res
